kuma/core/middleware.py

```diff
@@ -27,15 +27,25 @@ def process_request(self, request):
         prefixer = Prefixer(request)
         set_url_prefixer(prefixer)
         full_path = prefixer.fix(prefixer.shortened_path)
+        lang = request.GET.get('lang')
 
-        if 'lang' in request.GET:
+        if lang in dict(settings.LANGUAGES):
             # Blank out the locale so that we can set a new one. Remove lang
             # from the query params so we don't have an infinite loop.
             prefixer.locale = ''
             new_path = prefixer.fix(prefixer.shortened_path)
-            query = dict((smart_str(k), v) for
-                         k, v in request.GET.iteritems() if k != 'lang')
-            return HttpResponsePermanentRedirect(urlparams(new_path, **query))
+            query = dict((smart_str(k), v) for k, v in request.GET.iteritems() if k != 'lang')
+
+            # 'lang' is only used on the language selection dropdown. If this is
+            # present it is safe to set language preference for the current
+            # user.
+            if request.user.is_anonymous():
+                cookie = settings.LANGUAGE_COOKIE_NAME
+                request.session[cookie] = request.GET['lang']
```

Comment:
This sets a session key, not a cookie. We don't want to create sessions for anonymous users.  It also sets the language based on the URL, which is not what we want.